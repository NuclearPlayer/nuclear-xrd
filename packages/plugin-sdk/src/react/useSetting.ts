import { useEffect, useMemo, useRef, useState } from 'react';

import type { SettingsHost, SettingValue } from '../types/settings';

export const useSetting = <T extends SettingValue = SettingValue>(
  host: SettingsHost | undefined,
  id: string,
) => {
  const hostRef = useRef(host);
  hostRef.current = host;

  const [currentValue, setCurrentValue] = useState<T | undefined>(undefined);

  useEffect(() => {
    if (!hostRef.current) return;
    let unsubscribe: (() => void) | undefined;
    let isMounted = true;

    hostRef.current.get<T>(id).then((initialValue) => {
      if (!isMounted) return;
      setCurrentValue(initialValue);
      unsubscribe = hostRef.current?.subscribe<T>(id, (nextValue) =>
        setCurrentValue(nextValue),
      );
    });

    return () => {
      isMounted = false;
      if (unsubscribe) unsubscribe();
    };
  }, [id, hostRef]);

  const setValue = useMemo(
    () => (nextValue: T) => {
      if (!hostRef.current) return;
      void hostRef.current.set<T>(id, nextValue);
    },
    [id],
  );

  return [currentValue, setValue] as const;
};
