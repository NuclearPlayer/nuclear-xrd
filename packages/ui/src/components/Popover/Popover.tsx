import {
  Popover as HeadlessPopover,
  PopoverBackdrop,
  PopoverButton,
  PopoverPanel,
  PopoverPanelProps,
} from '@headlessui/react';
import { AnimatePresence, motion } from 'framer-motion';

import { cn } from '../../utils';

export type PopoverProps = {
  trigger: React.ReactNode;
  children: React.ReactNode;
  className?: string;
  anchor?: PopoverPanelProps['anchor'];
  backdrop?: boolean;
};

export const Popover = ({
  trigger,
  children,
  className,
  anchor,
  backdrop,
}: PopoverProps) => {
  return (
    <HeadlessPopover className={cn('absolute', className)}>
      {({ open }) => (
        <>
          <PopoverButton className="cursor-pointer">{trigger}</PopoverButton>
          <AnimatePresence>
            {open && (
              <>
                {backdrop && (
                  <PopoverBackdrop
                    transition
                    className="fixed inset-0 bg-black/20 transition duration-150 ease-out data-closed:opacity-0"
                  />
                )}
                <PopoverPanel
                  static
                  as={motion.div}
                  initial={{ opacity: 0, y: 4, scale: 0.98 }}
                  animate={{ opacity: 1, y: 8, scale: 1 }}
                  exit={{ opacity: 0, y: 4, scale: 0.98 }}
                  className={cn(
                    'text-foreground bg-primary border-border rounded border-2 px-4 py-2 leading-5 select-none',
                    {
                      ['translate-y-0']: anchor === 'bottom',
                      ['-translate-y-4']: anchor === 'top',
                      ['translate-x-2 -translate-y-2']: anchor === 'right',
                      ['-translate-x-2 -translate-y-2']: anchor === 'left',
                    },
                  )}
                  anchor={anchor}
                >
                  {children}
                </PopoverPanel>
              </>
            )}
          </AnimatePresence>
        </>
      )}
    </HeadlessPopover>
  );
};
