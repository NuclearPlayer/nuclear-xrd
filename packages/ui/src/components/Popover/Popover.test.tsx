import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { Button } from '..';
import { Popover } from './Popover';

describe('Popover', () => {
  it('(Snapshot) renders with all props', async () => {
    render(
      <Popover
        className="border-accent-red"
        anchor="right"
        trigger={<Button>Open Popover</Button>}
      >
        Popover Content
      </Popover>,
    );
    await userEvent.click(screen.getByText('Open Popover'));
    await screen.findByText('Popover Content');
    expect(document.body).toMatchSnapshot();
  });

  it('(Snapshot) renders with backdrop', async () => {
    render(
      <Popover trigger={<Button>Open Popover</Button>} backdrop>
        Popover Content
      </Popover>,
    );
    await userEvent.click(screen.getByText('Open Popover'));
    await screen.findByText('Popover Content');
    expect(document.body).toMatchSnapshot();
  });
});
