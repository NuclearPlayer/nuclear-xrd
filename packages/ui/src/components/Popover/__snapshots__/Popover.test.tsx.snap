// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Popover > (Snapshot) renders with all props 1`] = `
<body>
  <div>
    <div
      class="absolute border-accent-red"
      data-headlessui-state="open"
      data-open=""
    >
      <button
        aria-controls="headlessui-popover-panel-:r4:"
        aria-expanded="true"
        class="cursor-pointer"
        data-active=""
        data-headlessui-state="open active hover"
        data-hover=""
        data-open=""
        id="headlessui-popover-button-:r2:"
        type="button"
      >
        <button
          class="inline-flex cursor-pointer items-center rounded-md transition-all disabled:cursor-not-allowed disabled:opacity-50 text-foreground bg-primary border-border shadow-shadow hover:translate-x-shadow-x hover:translate-y-shadow-y border-2 hover:shadow-none h-10 px-4 py-2"
          data-headlessui-state="hover"
          data-hover=""
          type="submit"
        >
          Open Popover
        </button>
      </button>
      <button
        aria-hidden="true"
        data-headlessui-focus-guard="true"
        id="headlessui-focus-sentinel-:r3:"
        style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;"
        type="button"
      />
    </div>
  </div>
  <div
    id="headlessui-portal-root"
  >
    <div
      data-headlessui-portal=""
    >
      <button
        aria-hidden="true"
        data-headlessui-focus-guard="true"
        id="headlessui-focus-sentinel-before-:r4:"
        style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;"
        type="button"
      />
      <div
        animate="[object Object]"
        class="text-foreground bg-primary border-border rounded-md border-2 px-4 py-2 leading-5 select-none translate-x-2 -translate-y-2"
        exit="[object Object]"
        initial="[object Object]"
        style="overflow: auto; max-width: 0px; max-height: min(var(--anchor-max-height, 100vh), 0px);"
      >
        Popover Content
      </div>
      <button
        aria-hidden="true"
        data-headlessui-focus-guard="true"
        id="headlessui-focus-sentinel-after-:r4:"
        style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;"
        type="button"
      />
    </div>
  </div>
</body>
`;

exports[`Popover > (Snapshot) renders with backdrop 1`] = `
<body>
  <div>
    <div
      class="absolute"
      data-headlessui-state="open"
      data-open=""
    >
      <button
        aria-controls="headlessui-popover-panel-:rb:"
        aria-expanded="true"
        class="cursor-pointer"
        data-active=""
        data-headlessui-state="open active hover"
        data-hover=""
        data-open=""
        id="headlessui-popover-button-:r8:"
        type="button"
      >
        <button
          class="inline-flex cursor-pointer items-center rounded-md transition-all disabled:cursor-not-allowed disabled:opacity-50 text-foreground bg-primary border-border shadow-shadow hover:translate-x-shadow-x hover:translate-y-shadow-y border-2 hover:shadow-none h-10 px-4 py-2"
          data-headlessui-state="hover"
          data-hover=""
          type="submit"
        >
          Open Popover
        </button>
      </button>
      <button
        aria-hidden="true"
        data-headlessui-focus-guard="true"
        id="headlessui-focus-sentinel-:r9:"
        style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;"
        type="button"
      />
      <div
        aria-hidden="true"
        class="fixed inset-0 bg-black/20 transition duration-150 ease-out data-closed:opacity-0"
        data-closed=""
        data-enter=""
        data-headlessui-state="open"
        data-open=""
        data-transition=""
        id="headlessui-popover-backdrop-:ra:"
        style=""
      />
      <button
        aria-hidden="true"
        data-headlessui-focus-guard="true"
        id="headlessui-focus-sentinel-before-:rb:"
        style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;"
        type="button"
      />
      <div
        animate="[object Object]"
        class="text-foreground bg-primary border-border rounded-md border-2 px-4 py-2 leading-5 select-none"
        exit="[object Object]"
        initial="[object Object]"
      >
        Popover Content
      </div>
      <button
        aria-hidden="true"
        data-headlessui-focus-guard="true"
        id="headlessui-focus-sentinel-after-:rb:"
        style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;"
        type="button"
      />
    </div>
  </div>
</body>
`;
